### PR TITLE
CB-18654 - Update flowlog table to add flow end time

### PR DESCRIPTION
**Pull request by vivprabu**
Jira - https://jira.cloudera.com/browse/CB-18654

Added new column to flowlog table which will be updated when the status of the flow is being updated.

### CODE REVIEW

**Comment by bergerdenes** [comment]
please rebase to master

File: flow/src/main/java/com/sequenceiq/flow/domain/FlowLog.java [hunk 1]
```diff
@@ -116,6 +120,35 @@ public FlowLog(Long resourceId,
...
+        this.flowChainId = flowChainId;
+        this.flowTriggerUserCrn = flowTriggerUserCrn;
+        this.nextEvent = nextEvent;
+        this.payload = payload;
+        this.payloadJackson = payloadJackson;
+        this.payloadType = payloadType;
+        this.variables = variables;
+        this.variablesJackson = variablesJackson;
+        this.flowType = flowType;
+        this.currentState = currentState;
```
Comment (bergerdenes) [inline flow/src/main/java/com/sequenceiq/flow/domain/FlowLog.java:150]: Why not reuse the previous ctor?

File: flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java [hunk 2]
```diff
@@ -183,7 +184,12 @@ public void saveChain(String flowChainId, String parentFlowChainId, FlowTriggerE
     @Override
     public void updateLastFlowLogStatus(FlowLog lastFlowLog, boolean failureEvent) {
         StateStatus stateStatus = failureEvent ? StateStatus.FAILED : StateStatus.SUCCESSFUL;
-        flowLogRepository.updateLastLogStatusInFlow(lastFlowLog.getId(), stateStatus);
+        Long currentTime = getCurrentTime();
```
Comment (bergerdenes) [inline flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java:187]: `System.currentTimeMillis()` ?

File: flow/src/main/resources/schema/flow/20221005170000_CB-18654_add_endtime_to_flowlog.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+-- // CB-18654 Add Dlow Endtime to flowlog
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE IF EXISTS flowlog ADD COLUMN IF NOT EXISTS endtime int8;
```
Comment (bergerdenes) [inline flow/src/main/resources/schema/flow/20221005170000_CB-18654_add_endtime_to_flowlog.sql:4]: The undo command is not symmetric with this (i.e. the first `IF EXISTS`)

File: flow/src/test/resources/schema/flow_test/20221005170000_CB-18654_add_endtime_to_flowlog.sql [hunk 4]
```diff
@@ -0,0 +1,8 @@
+-- // CB-18654 Add Dlow Endtime to flowlog
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE IF EXISTS flowlog ADD COLUMN IF NOT EXISTS endtime int8;
+
+-- //@UNDO
+-- SQL to undo the change goes here.
+ALTER TABLE flowlog DROP COLUMN IF EXISTS endtime;
```
Comment (bergerdenes) [inline flow/src/test/resources/schema/flow_test/20221005170000_CB-18654_add_endtime_to_flowlog.sql:8]: same as above

Comment (bergerdenes) [reply flow/src/test/resources/schema/flow_test/20221005170000_CB-18654_add_endtime_to_flowlog.sql:8]: On the other hand: do you use this file in tests?

**Review by bergerdenes — CHANGES_REQUESTED** [review]
See my comments

File: flow/src/main/java/com/sequenceiq/flow/domain/FlowLog.java (hunk 1 above)
Comment (vivprabu) [reply flow/src/main/java/com/sequenceiq/flow/domain/FlowLog.java:150]: the previous constructor was being used and i didnt want to change any existing code

File: flow/src/main/resources/schema/flow/20221005170000_CB-18654_add_endtime_to_flowlog.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+-- // CB-18654 Add Dlow Endtime to flowlog
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply flow/src/main/resources/schema/flow/20221005170000_CB-18654_add_endtime_to_flowlog.sql:1]: done

File: flow/src/test/resources/schema/flow_test/20221005170000_CB-18654_add_endtime_to_flowlog.sql (hunk 4 above)
Comment (bergerdenes) [reply flow/src/test/resources/schema/flow_test/20221005170000_CB-18654_add_endtime_to_flowlog.sql]: ok

File: flow/src/main/java/com/sequenceiq/flow/domain/FlowLog.java (hunk 1 above)
Comment (bergerdenes) [reply flow/src/main/java/com/sequenceiq/flow/domain/FlowLog.java:150]: Call with this and set the new parameter only.

File: flow/src/main/resources/schema/flow/20221005170000_CB-18654_add_endtime_to_flowlog.sql [hunk 6]
```diff
@@ -0,0 +1,8 @@
+-- // CB-18654 Add Flow Endtime to flowlog
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE IF EXISTS flowlog ADD COLUMN IF NOT EXISTS endtime int8;
```
Comment (bergerdenes) [inline flow/src/main/resources/schema/flow/20221005170000_CB-18654_add_endtime_to_flowlog.sql:4]: I have checked into the code on how this column is used and find the naming misleading. It is the last update timestamp on a flow state.

File: flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java (hunk 2 above)
Comment (bergerdenes) [reply flow/src/main/java/com/sequenceiq/flow/service/flowlog/FlowLogDBService.java:187]: We have a nice `Clock` Spring service. Please look into that.

File: flow/src/test/java/com/sequenceiq/flow/service/flowlog/FlowLogDBServiceTest.java [hunk 7]
```diff
@@ -82,6 +84,7 @@ class FlowLogDBServiceTest {
             .build().toString();
 
     @InjectMocks
+    @Spy
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply flow/src/test/java/com/sequenceiq/flow/service/flowlog/FlowLogDBServiceTest.java:87]: This is an antipattern. Please use injectable services if you really need to test that.

**Review by bergerdenes — CHANGES_REQUESTED** [review]
See comments

**Review by bergerdenes — APPROVED** [review]

**Pull request merged**